pkg/server/endpoints/eventTracker.go

```diff
@@ -0,0 +1,264 @@
+ * increments each event's time, allowing every event to act on
+ * the time the event was inserted into eventTracker.
+ *
+ * The event's boundary Index is computed, and if it is below the
+ * number of initial polls, the event is polled without further
+ * analysis.
+ *
+ * If the boundary index is inside a defined boundary, the width
+ * of the boundary is computed and the event's position within
+ * the virtual hash table is computed from "hash(event) % width".
+ * As the hash and width are stable, an event will always remain
+ * in the same slot within a boundary.
+ *
+ * If the event's ticks (the events local sense of time) match
+ * the event's slot within the boundary, the event is added to
+ * the poll list.
+ */
+func (et *eventTracker) SelectEvents() []uint {
+	pollList := make([]uint, 0)
+	for event, _ := range et.events {
+		if et.events[event].ticks >= et.pollPeriods {
+			et.StopTracking(event)
+			continue
+		}
+		eventStats := et.events[event]
+		boundaryIndex := eventStats.polls - et.initialPolls
+		switch {
+		// before boundaries
+		case eventStats.polls < et.initialPolls:
+			pollList = append(pollList, event)
+			eventStats.polls++
+		// between boundaries
+		case boundaryIndex+1 < uint(len(et.boundaries)):
+			boundaryWidth := et.boundaries[1+boundaryIndex] - et.boundaries[boundaryIndex]
+			boundaryPosition := eventStats.hash % boundaryWidth
+			if eventStats.ticks == et.boundaries[boundaryIndex]+boundaryPosition {
+				pollList = append(pollList, event)
+				eventStats.polls++
+			}
+		// last boundary
+		case boundaryIndex < uint(len(et.boundaries)):
+			boundaryWidth := et.pollPeriods - et.boundaries[boundaryIndex]
+			boundaryPosition := eventStats.hash % boundaryWidth
+			if eventStats.ticks == et.boundaries[boundaryIndex]+boundaryPosition {
+				pollList = append(pollList, event)
+				eventStats.polls++
+			}
+		}
+		eventStats.ticks++
+	}
+	return pollList
+}
+
+/**
+ * Returns the count of events being tracked.
+ *
+ * @return the events being tracked.
+ */
+func (et *eventTracker) EventCount() uint {
+	return uint(len(et.events))
+}
+
+/**
+ * A hash function for uint.
+ *
+ * The slots within a boundary are conceptually a hash table, even
+ * though the hash table doesn't exist as an struct.  This means that
+ * each event being polled must distribute within the conceptual hash
+ * table evenly, or the conceptual hash table will only have entries in
+ * a subset of slots.
+ *
+ * This hashing algorithm is the modification of a number of algorithms
+ * previously found on the internet.  It avoids the factorization problem
+ * (even events only go into even slots) by repeatedly mixing high order
+ * bits into the low order bits ( h^h >> (number)).  The high order bits
+ * are primarily set by the low order bits repeatedly by multipliation with
+ * a number designed to mix bits deterministicly for better hash dispersion.
+ */
+func hash(event uint) uint {
```

Comment:
4_294_967_295 is max uint (4 billion).  At 800_000 a day, that would be 5_368 days, or about 14.6 years.

I imagine that in order to guard against this, we would have to dig into the database setup code that GORM creates.  That's because the database is the origin of the event's id, set through autoincrement.

Since Uber is skipping one for every entry, Uber would be burning through these at twice the 800_000 rate, so you have 7 years to address the issue.

Also, the odds are very good that your mysql implementation is using Integer, and not Uint, so you might have 1/2 of that.  Note that mysql (last I checked) had a very unusual way of handling the issue, it simply reused the max_int value over and over again, so you would get a strong failure (can't use a primary key for two entries).

To reset this you could shutdown the servers, delete the event tables, and relaunch.  There's nothing persisted about the next item to process, and so the system would start again at 0.

Reply:
`uint` is arch dependent. We only ship 64-bit artifacts so unless someone is building against 32-bit (unlikely), then the uint value itself will be 64-bits. That said, GORM (maybe now, and at least with the version we are using) uses 32-bit integer values for IDs of type `uint`. 

Ideally, SPIRE declare should declare its own model type using an explicit 64-bit type and then provide a migration to move the IDs to a 64-bit database type.